### DISCUSSION TITLE
Meassure Azure Function usage

### DISCUSSION THREAD

**Discussion by autosysops** [cloud-carbon-footprint/cloud-carbon-footprint#1255] [Feature suggestions]
When looking in the documentation for Azure functions (Consumption type) it talks about execution units being in the unit of GB-sec (https://learn.microsoft.com/en-us/azure/azure-functions/functions-consumption-costs?tabs=portal#consumption-plan-costs). From what I can see this unit can be read in the metrics for an Azure Function. There also seems to be a metric for the "memory working set". I believe with these two values combined it should be possible to figure out the execution time and memory usage.
The execution time could also be calculated by looking at the time different actions took which is also a metric.

I believe that for consuption type functions we should be able to take these execution seconds and combine them with a avarage cpu power usage to get some kind of estimate for the power usages of azure functions consumption type.

For the non consumption type it uses app service plans which I believe are being able to be calculated already (right?). And if not I believe there are metrics available for them to calculate this (I could try this if needed). Being able to measure the usage for azure functions both hosted and consumption type could really help in making better descissions when looking at the data.

**Comment by camcash17** [comment]
Hi @autosysops, thanks for bringing this up!

For some added context, as CCF reads over the Azure Consumption Management API (the current billing implementation), as it reads over a given response for Azure Function, when it sees that the pricing unit is GB-sec, it will get classified as a Memory usage type and go through the respective calculcation. (see [here](https://github.com/cloud-carbon-footprint/cloud-carbon-footprint/blob/ff2c408e673aa568bc220fabeb4b2325b59e81a2/packages/azure/src/lib/ConsumptionTypes.ts#L85).

What I would suggest is considering logic where if CCF faces the service name of "Azure Functions", even though the pricing unit is GB-Sec, we could add additional logic to classify the usage or calculate beyond memory.

Hope that makes sense!
